File: leetcode/137-single-number-ii.cpp
/*
 * Given an array of integers, every element appears three times except for one. Find that single one.

Note:
Your algorithm should have a linear runtime complexity. Could you implement it without using extra memory?
*/

class Solution {
public:
    int singleNumber(vector<int>& nums) {
        int result(0);
        for (auto i(31); i>=0; --i) {
            auto counter(0);
            for (auto num : nums) {
                if (((num >> i) & 1) == 1) {
                    ++counter;
                }
            }
            if (counter % 3 == 0) {
                result = result << 1;
            } else {
                result = (result << 1) | 1;
            }
        }
        return result;
    }
};

File: leetcode/137-single-number-ii.rb
# @param {Integer[]} nums
# @return {Integer}
def single_number(nums)
  result = 0
  (0..31).each do |i|
    counter = 0
    nums.each do |x|
      counter += 1 if ((x >> i) & 1) == 1
    end
    result |= (1 << i) if counter % 3 != 0
  end
  [result].pack('L').unpack('l')[0]
end

